### ISSUE TITLE
Setting min sample leaf for Quantile Regression Forest on Boston dataset

### ISSUE THREAD

**Issue by JChen2857** [scikit-garden/scikit-garden#50]
Hi,

I was playing around with the example Boston Data set and noticed if you change the parameter min_samples_leaf != 1, then the x_weights do not look like they are working properly and I run into the following error:

```Traceback (most recent call last):
  File "playaround.py", line 23, in <module>
    upper = np.concatenate((upper, rfqr.predict(X_test, quantile=98.5)))
  File "C:\Users\jchen\AppData\Local\Continuum\Anaconda2\lib\site-packages\skgarden\quantile\ensemble.py", line 143, in predict
    self.y_train_, quantile, weights, sorter)
  File "C:\Users\jchen\AppData\Local\Continuum\Anaconda2\lib\site-packages\skgarden\quantile\utils.py", line 69, in weighted_percentile
    total = sorted_cum_weights[-1]
IndexError: index -1 is out of bounds for axis 0 with size 0

Here is the sample Boston Dataset Code:

import matplotlib.pyplot as plt
import numpy as np
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from skgarden import RandomForestQuantileRegressor


boston = load_boston()
X, y = boston.data, boston.target
kf = KFold(n_splits=5, random_state=0)
rfqr = RandomForestQuantileRegressor(random_state=0, min_samples_split=10, n_estimators=1000,min_samples_leaf=3)

y_true_all = []
lower = []
upper = []

for train_index, test_index in kf.split(X):
	X_train, X_test, y_train, y_test = (X[train_index], X[test_index], y[train_index], y[test_index])
	rfqr.set_params(max_features=X_train.shape[1] // 3)
	rfqr.fit(X_train, y_train)
	y_true_all = np.concatenate((y_true_all, y_test))
	upper = np.concatenate((upper, rfqr.predict(X_test, quantile=98.5)))
	lower = np.concatenate((lower, rfqr.predict(X_test, quantile=2.5)))

interval = upper - lower
sort_ind = np.argsort(interval)
y_true_all = y_true_all[sort_ind]
upper = upper[sort_ind]
lower = lower[sort_ind]
mean = (upper + lower) / 2


# Center such that the mean of the prediction interval is at 0.0
y_true_all -= mean
upper -= mean
lower -= mean

plt.plot(y_true_all, "ro")
plt.fill_between(
    np.arange(len(upper)), lower, upper, alpha=0.2, color="r",
    label="Pred. interval")
plt.xlabel("Ordered samples.")
plt.ylabel("Values and prediction intervals.")
plt.xlim([0, 500])
plt.show()

Maybe I do not understand Quantile Regression Forests and the implementation is only suppose to have min_samples_leaf = 1, but that seems to contradict min_samples_split = 10 in the Boston Data Set example. Any help would be appreciated, thanks!

**Comment by JChen2857** [comment]
So the problem seems to be with the calculation of the weights for each observed data point (y_i's) in a leaf node. The weights are stored as an array in the fit method of BaseForestQuantileRegressor and a part of the code which calculates the weights (basically just the inverse of the number of data points in each leaf node, if the data point is in the leaf node) is using int division instead of float division. The only change made was:  

```self.y_weights_[i, y_ind] = (est_weights[y_ind] / np.sum(est_weights[y_ind]).astype(float))```

In this method:

```class BaseForestQuantileRegressor(ForestRegressor):
    def fit(self, X, y):
        """
        Build a forest from the training set (X, y).

        Parameters
        ----------
        X : array-like or sparse matrix, shape = [n_samples, n_features]
            The training input samples. Internally, it will be converted to
            ``dtype=np.float32`` and if a sparse matrix is provided
            to a sparse ``csc_matrix``.

        y : array-like, shape = [n_samples] or [n_samples, n_outputs]
            The target values (class labels) as integers or strings.

        sample_weight : array-like, shape = [n_samples] or None
            Sample weights. If None, then samples are equally weighted. Splits
            that would create child nodes with net zero or negative weight are
            ignored while searching for a split in each node. Splits are also
            ignored if they would result in any single class carrying a
            negative weight in either child node.

        check_input : boolean, (default=True)
            Allow to bypass several input checking.
            Don't use this parameter unless you know what you do.

        X_idx_sorted : array-like, shape = [n_samples, n_features], optional
            The indexes of the sorted training input samples. If many tree
            are grown on the same dataset, this allows the ordering to be
            cached between trees. If None, the data will be sorted here.
            Don't use this parameter unless you know what to do.

        Returns
        -------
        self : object
            Returns self.
        """
        # apply method requires X to be of dtype np.float32
        X, y = check_X_y(
            X, y, accept_sparse="csc", dtype=np.float32, multi_output=False)
        super(BaseForestQuantileRegressor, self).fit(X, y)

        self.y_train_ = y
        self.y_train_leaves_ = -np.ones((self.n_estimators, len(y)), dtype=np.int32)
        self.y_weights_ = np.zeros_like((self.y_train_leaves_), dtype=np.float32)
        # print 'Initialized array of y weights',self.y_weights_
        for i, est in enumerate(self.estimators_):
            if self.bootstrap:
                bootstrap_indices = generate_sample_indices(
                    est.random_state, len(y))
            else:
                bootstrap_indices = np.arange(len(y))

            est_weights = np.bincount(bootstrap_indices, minlength=len(y))
            # print 'estimator_weights',est_weights
            y_train_leaves = est.y_train_leaves_
            for curr_leaf in np.unique(y_train_leaves):
                y_ind = y_train_leaves == curr_leaf
                self.y_weights_[i, y_ind] = (
                    est_weights[y_ind] / np.sum(est_weights[y_ind]).astype(float))
                sum_test = np.sum(est_weights[y_ind])

            self.y_train_leaves_[i, bootstrap_indices] = y_train_leaves[bootstrap_indices]
            self.y_train_leaves_[i, bootstrap_indices] = y_train_leaves[bootstrap_indices]
        return self
```
Furthermore, while examining the code I was confused on how weights were averaged over each decision tree. In the predict method of BaseForestQuantileRegressor the quantiles are predicted by taking all relevant leaf nodes over the number of trees used in the forest and is summed together in the  weights = x_weights.sum(axis = 0). However this sum is not normalized over the number of trees before being used in the weighted_percentile function. If someone can please shed some light over this question (maybe @MechCoder MechCoder?).  The method in question:

``` def predict(self, X, quantile=None):
        """
        Predict regression value for X.

        Parameters
        ----------
        X : array-like or sparse matrix of shape = [n_samples, n_features]
            The input samples. Internally, it will be converted to
            ``dtype=np.float32`` and if a sparse matrix is provided
            to a sparse ``csr_matrix``.

        quantile : int, optional
            Value ranging from 0 to 100. By default, the mean is returned.

        check_input : boolean, (default=True)
            Allow to bypass several input checking.
            Don't use this parameter unless you know what you do.

        Returns
        -------
        y : array of shape = [n_samples]
            If quantile is set to None, then return E(Y | X). Else return
            y such that F(Y=y | x) = quantile.
        """
        # apply method requires X to be of dtype np.float32
        X = check_array(X, dtype=np.float32, accept_sparse="csc")
        if quantile is None:
            return super(BaseForestQuantileRegressor, self).predict(X)

        sorter = np.argsort(self.y_train_)
        X_leaves = self.apply(X)
        weights = np.zeros((X.shape[0], len(self.y_train_)))
        quantiles = np.zeros((X.shape[0]))
        for i, x_leaf in enumerate(X_leaves):
            mask = self.y_train_leaves_ != np.expand_dims(x_leaf, 1)
            x_weights = ma.masked_array(self.y_weights_, mask)
            weights = x_weights.sum(axis=0)
            quantiles[i] = weighted_percentile(
                self.y_train_, quantile, weights, sorter)
        return quantiles

**Comment by MechCoder** [comment]
1. I have tested it only with Python3 where the division works as expected. Feel free to send a PR that fixes this.
2. It does not matter if the weights are scaled such that they lie in-between [0, 1] or not since the `weighted_percentile` implementation takes care of that internally.

Hope this answers your questions!

**Comment by JChen2857** [comment]
Got it, thanks. Should be fixed in #51 .
